packages/backend-function/src/runtime/get_amplify_clients_configuration.ts

```diff
@@ -0,0 +1,156 @@
+    typeof env.AMPLIFY_DATA_MODEL_INTROSPECTION_SCHEMA_BUCKET_NAME ===
+      'string' &&
+    typeof env.AMPLIFY_DATA_MODEL_INTROSPECTION_SCHEMA_KEY === 'string' &&
+    typeof env.AWS_ACCESS_KEY_ID === 'string' &&
+    typeof env.AWS_SECRET_ACCESS_KEY === 'string' &&
+    typeof env.AWS_SESSION_TOKEN === 'string' &&
+    typeof env.AWS_REGION === 'string' &&
+    typeof env.AMPLIFY_DATA_GRAPHQL_ENDPOINT === 'string'
+  );
+};
+
+const getResourceConfig = (
+  env: DataClientEnv,
+  modelIntrospectionSchema: object
+) => {
+  return {
+    API: {
+      GraphQL: {
+        endpoint: env.AMPLIFY_DATA_GRAPHQL_ENDPOINT,
+        region: env.AWS_REGION,
+        defaultAuthMode: 'iam' as const,
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        modelIntrospection: modelIntrospectionSchema as any,
+      },
+    },
+  };
+};
+
+const getLibraryOptions = (env: DataClientEnv) => {
+  return {
+    Auth: {
+      credentialsProvider: {
+        getCredentialsAndIdentityId: async () => ({
+          credentials: {
+            accessKeyId: env.AWS_ACCESS_KEY_ID,
+            secretAccessKey: env.AWS_SECRET_ACCESS_KEY,
+            sessionToken: env.AWS_SESSION_TOKEN,
+          },
+        }),
+        clearCredentialsAndIdentityId: () => {
+          /* noop */
+        },
+      },
+    },
+  };
+};
+
+type InvalidConfig = unknown & {
+  invalidType: 'This function needs to be granted `authorization((allow) => [allow.resource(fcn)])` on the data schema.';
+};
+
+type DataClientError = {
+  resourceConfig: InvalidConfig;
+  libraryOptions: InvalidConfig;
+};
+
+type DataClientConfig = {
+  resourceConfig: ReturnType<typeof getResourceConfig>;
+  libraryOptions: ReturnType<typeof getLibraryOptions>;
+};
+
+type DataClientReturn<T> = T extends DataClientEnv
+  ? DataClientConfig
+  : DataClientError;
+
+/**
+ * Generate the `resourceConfig` and `libraryOptions` need to configure
+ * Amplify for the data client in a lambda.
+ *
+ * Your function needs to be granted resource access on your schema for this to work
+ * `a.schema(...).authorization((allow) => [a.resource(myFunction)])`
+ * @param env - The environment variables for the data client
+ * @returns An object containing the `resourceConfig` and `libraryOptions`
+ */
+export const getAmplifyClientsConfiguration = async <T>(
+  env: T
+): Promise<DataClientReturn<T>> => {
+  const s3Client = new S3Client();
+  return getAmplifyClientsConfigurationRetriever(env, s3Client);
+};
+
+/**
+ * The implementation of `getAmplifyClientsConfiguration` that allows you to pass in an S3 client for testing purposes
+ * @param env - The environment variables for the data client
+ * @param s3Client - The S3 client to use for retrieving the schema
+ * @returns An object containing the `resourceConfig` and `libraryOptions`
+ */
+export const getAmplifyClientsConfigurationRetriever = async <T>(
+  env: T,
+  s3Client: S3Client
+): Promise<DataClientReturn<T>> => {
```

Comment:
This can be single function, you can

```
export const getAmplifyClientsConfigurationRetriever = async <T>(
  env: T,
  s3Client = new S3Client()
): Promise<DataClientReturn<T>> => {
```

OR do what ssm is doing here https://github.com/aws-amplify/amplify-backend/blob/9fd064261b1a8ee3c9f544885f68b128a82564db/packages/backend-function/src/lambda-shims/resolve_ssm_params.ts#L12

Reply:
Got it. Making the change. Should the export function name be prefixed with internal. eg `internalGetAmplifyClientConfiguration` or similar?

Reply:
yes if customer is not supposed to call it themselves